File: php/test/WhatsAppTemplateGetV2PaginationEntityTest.php
<?php
declare(strict_types=1);

// WhatsAppTemplateGetV2Pagination entity test

require_once __DIR__ . '/../lmwhatsapp_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class WhatsAppTemplateGetV2PaginationEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = LmWhatsappSDK::test(null, null);
        $ent = $testsdk->WhatsAppTemplateGetV2Pagination(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = whats_app_template_get_v2_pagination_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "whats_app_template_get_v2_pagination." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set LMWHATSAPP_TEST_WHATS_APP_TEMPLATE_GET_V__PAGINATION_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $whats_app_template_get_v2_pagination_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.whats_app_template_get_v2_pagination")));
        $whats_app_template_get_v2_pagination_ref01_data = null;
        if (count($whats_app_template_get_v2_pagination_ref01_data_raw) > 0) {
            $whats_app_template_get_v2_pagination_ref01_data = Helpers::to_map($whats_app_template_get_v2_pagination_ref01_data_raw[0][1]);
        }

        // LOAD
        $whats_app_template_get_v2_pagination_ref01_ent = $client->WhatsAppTemplateGetV2Pagination(null);
        $whats_app_template_get_v2_pagination_ref01_match_dt0 = [];
        $whats_app_template_get_v2_pagination_ref01_data_dt0_loaded = $whats_app_template_get_v2_pagination_ref01_ent->load($whats_app_template_get_v2_pagination_ref01_match_dt0, null);
        $this->assertNotNull($whats_app_template_get_v2_pagination_ref01_data_dt0_loaded);

    }
}

function whats_app_template_get_v2_pagination_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/whats_app_template_get_v2_pagination/WhatsAppTemplateGetV2PaginationTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = LmWhatsappSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["whats_app_template_get_v2_pagination01", "whats_app_template_get_v2_pagination02", "whats_app_template_get_v2_pagination03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("LMWHATSAPP_TEST_WHATS_APP_TEMPLATE_GET_V__PAGINATION_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "LMWHATSAPP_TEST_WHATS_APP_TEMPLATE_GET_V__PAGINATION_ENTID" => $idmap,
        "LMWHATSAPP_TEST_LIVE" => "FALSE",
        "LMWHATSAPP_TEST_EXPLAIN" => "FALSE",
        "LMWHATSAPP_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["LMWHATSAPP_TEST_WHATS_APP_TEMPLATE_GET_V__PAGINATION_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["LMWHATSAPP_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["LMWHATSAPP_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new LmWhatsappSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["LMWHATSAPP_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["LMWHATSAPP_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/whats_app_template_get_v2_pagination_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/lm-whatsapp-sdk/go"
	"github.com/voxgig-sdk/lm-whatsapp-sdk/go/core"

	vs "github.com/voxgig-sdk/lm-whatsapp-sdk/go/utility/struct"
)

func TestWhatsAppTemplateGetV2PaginationEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.WhatsAppTemplateGetV2Pagination(nil)
		if ent == nil {
			t.Fatal("expected non-nil WhatsAppTemplateGetV2PaginationEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := whats_app_template_get_v2_paginationBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "whats_app_template_get_v2_pagination." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set LM_WHATSAPP_TEST_WHATS_APP_TEMPLATE_GET_V2_PAGINATION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		whatsAppTemplateGetV2PaginationRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath(setup.data, "existing.whats_app_template_get_v2_pagination")))
		var whatsAppTemplateGetV2PaginationRef01Data map[string]any
		if len(whatsAppTemplateGetV2PaginationRef01DataRaw) > 0 {
			whatsAppTemplateGetV2PaginationRef01Data = core.ToMapAny(whatsAppTemplateGetV2PaginationRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = whatsAppTemplateGetV2PaginationRef01Data

		// LOAD
		whatsAppTemplateGetV2PaginationRef01Ent := client.WhatsAppTemplateGetV2Pagination(nil)
		whatsAppTemplateGetV2PaginationRef01MatchDt0 := map[string]any{}
		whatsAppTemplateGetV2PaginationRef01DataDt0Loaded, err := whatsAppTemplateGetV2PaginationRef01Ent.Load(whatsAppTemplateGetV2PaginationRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if whatsAppTemplateGetV2PaginationRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func whats_app_template_get_v2_paginationBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "whats_app_template_get_v2_pagination", "WhatsAppTemplateGetV2PaginationTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read whats_app_template_get_v2_pagination test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse whats_app_template_get_v2_pagination test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"whats_app_template_get_v2_pagination01", "whats_app_template_get_v2_pagination02", "whats_app_template_get_v2_pagination03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("LM_WHATSAPP_TEST_WHATS_APP_TEMPLATE_GET_V2_PAGINATION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"LM_WHATSAPP_TEST_WHATS_APP_TEMPLATE_GET_V2_PAGINATION_ENTID": idmap,
		"LM_WHATSAPP_TEST_LIVE":      "FALSE",
		"LM_WHATSAPP_TEST_EXPLAIN":   "FALSE",
		"LM_WHATSAPP_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["LM_WHATSAPP_TEST_WHATS_APP_TEMPLATE_GET_V2_PAGINATION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["LM_WHATSAPP_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["LM_WHATSAPP_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewLmWhatsappSDK(core.ToMapAny(mergedOpts))
	}

	live := env["LM_WHATSAPP_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["LM_WHATSAPP_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
